Add snap color option for osu!mania

JustusFT opened the pull request:
Ref: https://github.com/ppy/osu/issues/8583

Color-coded notes can be enabled in the options.
The snap colors are the same as the editor's colors.

### Screenshots:
Colored hit notes:
![osu_2021-04-24_16-47-32](https://user-images.githubusercontent.com/7854367/115953395-cef92080-a51d-11eb-91c9-c97e7f026728.jpg)

Colored hold notes:
![osu_2021-04-24_16-46-42](https://user-images.githubusercontent.com/7854367/115953404-d7e9f200-a51d-11eb-87f2-8157fd8d0dba.jpg)

JustusFT committed:
Add snap color option for osu!mania

JustusFT reviewed (commented):
Some notes on the PR

bdach commented on osu.Game.Rulesets.Mania/UI/ManiaColourCode.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace osu.Game.Rulesets.Mania.UI
+{
+    public enum ManiaColourCode
+    {
+        Off,
+        On
+    }
```
Not sure why this is a thing. What you have there is a glorified `bool`. I don't think there will ever be any more options than these.

bdach reviewed (changes requested):
Very superficial review, but the core design of this PR is incorrect, so I'm not looking into the details at all.

Please split commits better. This should not be a single-commit PR.

JustusFT commented on osu.Game.Rulesets.Mania/Objects/Note.cs:
```diff
@@ -12,5 +17,79 @@ namespace osu.Game.Rulesets.Mania.Objects
+
+        public Note(IBeatmap beatmap) : base()
+        {
+            Beatmap = beatmap;
```
Noted, my original approach was having the beatmap injected in the `DrawableNote` but my thought the better location would be closer to `Note` so I could separate the data (`Snap`) from the visuals (`Colour`). I admit that this approach ended up looking too rough

I'll follow your suggestion redo the beatmap part to use DI

peppy replied:
You should either be using an `IBeatSnapProvider`, similar to the editor, if anything.

JustusFT replied:
Hi @peppy, I took a look into `IBeatSnapProvider` but from my understanding it finds the closest hit time given the current snap. I want to do the inverse (find the closest snap given the hit time) so I'm not sure I could use it. So instead I added a new class (`SnapFinder.cs`) to do this

peppy replied:
I think we will eventually want the same logic in the editor, but that's fine for now. Can move or combine during review.

JustusFT committed:
Remove beatmap argument in Note

JustusFT committed:
Move color snap logic from Note to DrawableNote

JustusFT committed:
Remove ManiaColourCode in favor for boolean

JustusFT committed:
Rename ColourCode to ColourCodedNotes

JustusFT committed:
Move snapping logic into SnapFinder

JustusFT committed:
Fix visual tests missing dependency for ColourCodedNotes

JustusFT committed:
Fix cake errors

JustusFT commented:
Hi @bdach, I followed your suggestions and reverted the constructor for `Note`. It's using using DI now.

The injected snapping logic can be found in `SnapFinder.cs`.

peppy commented on osu.Game.Rulesets.Mania/UI/DrawableManiaRuleset.cs:
```diff
@@ -45,12 +46,18 @@ public class DrawableManiaRuleset : DrawableScrollingRuleset<ManiaHitObject>
 
         protected new ManiaRulesetConfigManager Config => (ManiaRulesetConfigManager)base.Config;
 
+        [Cached]
+        protected readonly Bindable<bool> ConfigColourCodedNotes = new Bindable<bool>();
```
For sure. Always work from example - most conventions are already used somewhere else. For instance, snaking in slider bodies: https://github.com/ppy/osu/blob/7654df94f6f37b8382be7dfcb4f674e03bd35427/osu.Game.Rulesets.Osu/Skinning/Default/PlaySliderBody.cs

JustusFT commented on osu.Game.Rulesets.Mania/Utils/SnapFinder.cs:
```diff
@@ -0,0 +1,76 @@
+        {
+            TimingControlPoint currentTimingPoint = beatmap.ControlPointInfo.TimingPointAt(hitObject.StartTime);
+            double startTime = currentTimingPoint.Time;
+            double secondsPerFourCounts = currentTimingPoint.BeatLength * 4;
```
The tempo of a song is affected by the BPM value as well as type of beat that is used to measure the BPM. It seems osu measures the BPM in quavers (quarter notes). So for 100bpm, the duration of a quaver is the same whether it is a `SimpleQuadruple` or `SimpleTriple`.

What I'm doing here is converting from quavers per minute to semibreves (whole notes) per second. Then I can divide the whole note to find the snap

Naxesss replied:
Multiplying the beat length by 4 here does seem odd, as all we're looking for is the snap within a beat.

Right now we first take the modulus of 4 beats, then the modulus of a fraction of a beat. The former step is redundant, because the latter step is a smaller multiple of the beat length. I think it would be less confusing for readers of the code if we didn't do the former step at all, because it makes it seem like the structure of beats matter.

I'd suggest simply passing `hitObject.StartTime - currentTimingPoint.Time` to `almostDivisibleBy`, which yields the same result as far as I've tested.

JustusFT replied:
My latest commit already removed the * 4 multiplication because it's redundant, but yeah you're right, I don't need the modulus of 4 beats either! Looks like I got locked into thinking in whole notes while working on this

JustusFT committed:
Simplify FindSnap method

JustusFT committed:
Move SnapFinder from mania ruleset to osu.Game

JustusFT committed:
Simplify DrawableNote

JustusFT committed:
Remove caching for ConfigColourCodedNotes

JustusFT committed:
Add xmldoc to SnapFinder

JustusFT committed:
Fix cake errors

JustusFT commented:
Hi @bdach and @peppy, thanks for your inputs. I have refactored based on your suggestions and also made `SnapFinder` a bit cleaner.

bdach committed:
Use null-permitting BDL to reduce number of fields

bdach committed:
Reduce casting

bdach committed:
Rename {Snap -> BeatDivisor}Finder

bdach committed:
Use existing snap list from BindableBeatDivisor

bdach commented:
I have made some follow-up changes. There is still a distinct lack of tests in this PR, but I don't believe I have any major problem remaining with the rest of this code.

@JustusFT Please let me know if you are willing to attempt to add testing, both of the `BeatDivisorFinder` (previously `SnapFinder` - renamed as I believe the new name to be better), and the visual part (simple visual test scene toggling the setting on and off will suffice).

JustusFT committed:
Add tests for BeatDivisorFinder

JustusFT commented:
@bdach Yes, I will add some tests for `BeatDivisorFinder` as well as some visual testing :+1:

peppy commented on osu.Game.Rulesets.Mania/ManiaSettingsSubsection.cs:
```diff
@@ -37,6 +37,11 @@ private void load()
                 },
+                new SettingsCheckbox
+                {
+                    LabelText = "Colour-coded notes",
```
I'd go with something like "Timing-based note colouring". As it stands it's ambiguous as to what they are colour-coded by (for instance notes can also be colour-coded by the column they are in).

frenzibyte commented on osu.Game.Rulesets.Mania/Objects/Drawables/DrawableNote.cs:
```diff
@@ -73,5 +100,12 @@ public virtual bool OnPressed(ManiaAction action)
+
+        private void updateSnapColour()
+        {
+            Colour = configColourCodedNotes.Value
```
Sounds like you would want to set this to `AccentColour.Value` instead, rather than the actual drawable colour.

peppy replied:
Have you tested that this doesn't change any behaviours?

frenzibyte replied:
I haven't but thought it's the correct direction for colouring drawable hit objects properly, rather than going for `Drawable.Colour`.

JustusFT replied:
I just tried it out, right now `AccentColour` doesn't affect the notes of the head and tail (`DefaultNotePiece`) but does affect with the body of a hold note (`DefaultBodyPiece`)

I could change `DefaultNotePiece` to use `AccentColour` but it will also affect the column colors of the hold notes which are currently set in `Stage.cs` ([this code](https://github.com/ppy/osu/blob/master/osu.Game.Rulesets.Mania/UI/Stage.cs#L40-L45)). Maybe we can add a second accent color value so we change them separately?

peppy commented on osu.Game.Rulesets.Mania/Objects/Drawables/DrawableNote.cs:
```diff
@@ -34,6 +47,20 @@ public DrawableNote(Note hitObject)
+            {
+                rulesetConfig?.BindWith(ManiaRulesetSetting.ColourCodedNotes, configColourCodedNotes);
+
+                HitObject.StartTimeBindable.BindValueChanged(_ => snap.Value = beatDivisorFinder.FindDivisor(HitObject), true);
+
+                snap.BindValueChanged(_ => updateSnapColour(), true);
+                configColourCodedNotes.BindValueChanged(_ => updateSnapColour());
```
Putting `BindValueChanged` hooks inside a BDL load method is generally frowned upon. Probably safest to move these to `LoadComplete`.

peppy replied:
Also just for legibility, I'd move the `BindValueChanged` with the `true` specification to be the last one in the method. That's how we do it everywhere else.

bdach replied:
this one is on me, not sure how I missed the callbacks.

JustusFT committed:
Move BindValueChanged hooks to LoadComplete()

JustusFT committed:
Rename ColourCodedNotes to TimingBasedNoteColouring

JustusFT committed:
Add visual tests for timing based note coloring

JustusFT committed:
Fix cake errors

JustusFT committed:
Merge branch 'master' of https://github.com/ppy/osu into justusft/mania-color-snap

JustusFT committed:
Replace BeatDivisorFinder with GetClosestBeatDivisor

JustusFT committed:
Fix GetClosestBeatDivisor returning the wrong divisor

JustusFT committed:
Remove unneeded test data

JustusFT commented:
Hi @ppy, I have made some commits to address your review.

@bdach I removed `BeatDivisorFinder` in favor of the new `ClosestBeatDivisor` method, so there's no more need to add tests to that. I also added a visual test for the note colors. Here's a screenshot:

![Screenshot2021-05-0114:21:26](https://user-images.githubusercontent.com/7854367/116773880-20a82a80-aa8b-11eb-8559-2be01cd4f3bd.png)

peppy committed:
Tidy up implementation in `DrawableNote`

peppy committed:
Move caching of `IBeatmap` to base `DrawableRuleset`

peppy commented:
I've tidied things up a bit. Should be pretty good for an initial implementation now.

bdach reviewed (approved):
Good enough for me, at least for an initial implementation.

peppy commented on osu.Game/Rulesets/UI/DrawableRuleset.cs:
```diff
@@ -85,6 +85,7 @@ internal bool FrameStablePlayback
         /// <summary>
         /// The beatmap.
         /// </summary>
+        [Cached(typeof(IBeatmap))]
```
I wasn't sure about this, but what part makes it scary?

bdach replied:
I don't think it should break anything, I'm mostly worried about future PRs using it too much for things that don't/shouldn't require the entire beatmap.
